doc/decisions/characters.md

```diff
@@ -0,0 +1,39 @@
+# Characters
+
+## Problem
+
+Users must be safe from interpretations of plugins.
+`keySetBaseName("@BASE64abc")` should not suddenly turn the name to something else (given a base64 plugin for names).
+Nor `keySetBaseName("#123")` should suddenly turn some keys to an array.
+
+It is not possible to allow all of following:
+
+- have any characters in key name parts for applications
+- being able to encode information in names
+
+## Constraints
+
+## Assumptions
+
+## Considered Alternatives
+
+## Decision
+
+Decide in favour of key name parts with any characters.
+
+For arrays there is actually not an conflict because an array is only an array if the meta-data `array` is appended to the direct parent key.
+See [array](array.md). So `#` does not need to be escaped.
+
+For extensibility of Elektra, however, at least one character needs a special meaning.
```

Comment:
You tell us please :wink:

Which kind of extensibility in key names is important for you?

Reply:
To me it is not about extensibility of key names. Its about giving storage plugins (and possibly other components) breathing room.

I also can't answer the question of "special meaning", because I can't read your mind...

Does "special meaning" mean:
- You can't use it in `keyNew`
- You can't use it in `keySetBaseName`
- You can use it but it implies certain things in certain contexts (like `#` arrays)
- Do I have to escape it? If so, when and where and how?
- Something else

I don't want to know the actual meaning, I want to know the general effects it has.

Reply:
> To me it is not about extensibility of key names. Its about giving storage plugins (and possibly other components) breathing room.

A storage plugin has full control over every detail of the syntax and can produce every possible KeySet and key part names can have every possible sequence of characters (except NULL bytes, I wrote about this in ). How can you give more breathing room?

> I also can't answer the question of "special meaning", because I can't read your mind...

I removed the decision: e53409d4e99ff4d599c6521b2cca5e3377a51839

Reply:
> A storage plugin has full control over every detail of the syntax [...]

Correct in general, but most plugins want to adhere to the pre-defined spec of a standard format (JSON, YAML, TOML).
Therefore, most plugins are quite limited in what syntax they can use.

> [...] and can produce every possible KeySet [...]

This may sound like it gives flexibility, but it actually is a huge constraint.

To make a storage plugin for a standard format useful, it should produce a `KeySet` whose structure somewhat reflects that of the storage file. This one of the goals you wrote. At the same time, transforming a `KeySet` into a storage format should also retain the structure as far as possible.

So while a storage plugin can in theory produce any `KeySet`. There has to be a well-defined relation between the storage formats structure and a `KeySet`. To make the plugin useful, this relation should be easy to understand and result in as few unexpected transformations as possible.

> [...] and key part names can have every possible sequence of characters [...]

Again, this may sound like flexibility, but from the point of a storage plugin it only means trouble.

First of all, most (modern) formats only support UTF-8 (or are defined based on Unicode). So we already have a problem. We either have to encode non-UTF-8 bytes somehow (such that they can be decoded) or we can't support everything a `KeySet` supports. In some formats encoding could be done with escape sequences, but not all formats (e.g. JSON) support those. So if we want to adhere to the spec, we need to find another way.

Note: The problem is hard to explain without an example, so I will use JSON.

The expected relation between a JSON object and a `KeySet` is that we use the JSON field names as key name parts. That means every JSON field should correspond to a key name part. But in a JSON there is nothing other than fields. So there is very little flexbility.

> [...] How can you give more breathing room?

At the risk of repeating myself: Restricting key names allows storage plugins to encode things that otherwise wouldn't be possible.

Basically, be restricting key names we create a hole in the structure of `KeySet`s that can be mapped to the parts of the structure that do not fit a storage format.

Every (general) storage plugin must be able to store keys with values, so if we can safely map all of a `KeySet`s semantics onto key-value pairs, any plugin can store `KeySet`s without loss of information. Of course it is already possible to find such a mapping. But what isn't possible is to find a mapping which also keeps all existing leaf key-value pairs in a `KeySet` unchanged.

Right now `KeySet`s do not leave any breathing room. **Every** random sequence of bytes corresponds to a key name, if you add a namespace and a zero-byte at the front and a zero-byte at the end. If you have a format like JSON that is just a single tree, every path in that in tree must correspond the key name, there is just no other way.